File: sm64py/aim.py
"""Turn the upper body toward what the player is aiming at.

This is docs/aim.md's procedural layer, and only that layer: the clips still
say how an attack moves, and this says which way the motion is pointed. The
authored animation is never overridden, because the bone this writes to --
AIM_TORSO, inserted by tools/aim_rig.py -- carries no keyframes at all. It sits
between the skeleton root and everything above the hips, so rotating it turns
the spine, the head, the arms and the cape as one piece while the clips
underneath keep playing whatever they were playing.

What that buys is the thing the Hero could not do before: aim somewhere other
than where he is walking. The legs run in the direction he is moving, the torso
turns toward the crosshair, and the two are no longer the same number.

Three rules shape it, all from the doc:

**The torso has limits.** It twists to `yaw_limit` and no further. Past that
the character has to turn his feet, which is what `body_turn` asks for -- it
returns the excess rather than turning him, since who is allowed to write to a
character's facing is the caller's business and not this file's.

**It does not snap.** The torso is sprung toward the aim rather than assigned
to it, and how fast is a per-weapon number: docs/aim.md's "Pistol fast, Rifle
fast, Cannon slow" is the `response` field of a profile. A hard mount that
whipped instantly to the crosshair would read as the camera dragging a doll
around rather than as a character looking.

**Melee gives it up as it commits.** `tracking` scales the whole correction, so
an attack can track hard during the windup and not at all once the blade is
moving. See `melee_tracking`.

The rest of the doc -- distributing the turn across spine and shoulders, aim
offset poses, left-hand IK -- needs a skeleton whose arms are parented to its
chest, which the Hero's export is not. tools/aim_rig.py says why at length.
"""

import math

from .math_util import atan2s, degrees_to_s16, s16, s16_to_degrees

# -- profiles -----------------------------------------------------------------


class AimProfile:
    """How one weapon aims. docs/aim.md's "part of weapon feel"."""

    def __init__(self, yaw_limit=60.0, pitch_min=-45.0, pitch_max=60.0,
                 response=0.12, pitch_share=0.55, comfort_yaw=20.0,
                 turn_rate=6.0):
        # How far the torso twists before the feet have to help, in degrees.
        self.yaw_limit = yaw_limit
        # Down and up. Aiming up leans him back, so up is the larger number:
        # a body bends further backwards from the hips than it folds forwards
        # over them without the legs joining in.
        self.pitch_min = pitch_min
        self.pitch_max = pitch_max
        # Spring time in seconds -- smaller is snappier. A rifle is fast, a
        # cannon is slow, and that difference is most of what the two feel like.
        self.response = response
        # How much of the aim's elevation the torso actually takes. Full pitch
        # rotates the whole upper body by the angle of the shot, which for a
        # steep one lays him flat; the rest of the angle belongs to the arms,
        # in a clip or an aim-offset pose that does not exist yet.
        self.pitch_share = pitch_share
        # Standing still, he squares up until the twist is back inside this.
        # Without it he would stand at a 59-degree twist indefinitely, which
        # the limit permits and no person does.
        self.comfort_yaw = comfort_yaw
        # How fast the feet come round, in turns per second at full deflection.
        self.turn_rate = turn_rate


# The Hero has no weapons yet, so there is one profile and it is a rifle's:
# quick, because everything about the camera it is attached to is quick.
DEFAULT = AimProfile()


# -- melee commitment ---------------------------------------------------------

# docs/aim.md's tracking curve, as (normalised time, tracking) knees. The
# attack steers freely while he is winding up, keeps some of it as the swing
# starts, almost none once the blade is live, and none at all on the recovery.
# Interpolated between the knees rather than stepped, so the correction bleeds
# away over the swing instead of switching off on one frame.
MELEE_TRACKING = ((0.00, 1.00), (0.25, 0.60), (0.55, 0.15), (0.75, 0.00))


def melee_tracking(normalised_time, curve=MELEE_TRACKING):
    """How much of the aim an attack still follows, this far into itself."""
    t = min(max(normalised_time, 0.0), 1.0)
    previous_time, previous_value = curve[0]
    if t <= previous_time:
        return previous_value
    for time, value in curve[1:]:
        if t <= time:
            span = time - previous_time
            share = (t - previous_time) / span if span else 1.0
            return previous_value + (value - previous_value) * share
        previous_time, previous_value = time, value
    return previous_value


# -- the controller -----------------------------------------------------------


def _spring(current, velocity, target, smooth_time, dt):
    """One step of a critically damped spring. See camera.smooth_damp.

    Copied in shape rather than imported so this module does not depend on the
    camera; it is the same algebra without the maximum-speed clamp.
    """
    smooth_time = max(smooth_time, 1e-4)
    omega = 2.0 / smooth_time
    x = omega * dt
    decay = 1.0 / (1.0 + x + 0.48 * x * x + 0.235 * x * x * x)
    change = current - target
    temp = (velocity + omega * change) * dt
    velocity = (velocity - omega * temp) * decay
    return target + (change + temp) * decay, velocity


class AimController:
    """Where the upper body is pointed, and what that costs the feet.

    Fed a world-space direction, it produces the character-local yaw and pitch
    of that direction, springs the torso toward them within the profile's
    limits, and writes the result to the AIM_TORSO joint.
    """

    def __init__(self, joint, profile=DEFAULT):
        # A Panda3D NodePath from Actor.controlJoint, or None -- a character
        # whose skeleton has no pivot still answers every question below, it
        # simply has nothing to draw the answer on.
        self.joint = joint
        self.profile = profile

        self.target_yaw = 0.0           # where the aim is, character-local
        self.target_pitch = 0.0
        self.yaw = 0.0                  # where the torso actually is
        self.pitch = 0.0
        self._yaw_velocity = 0.0
        self._pitch_velocity = 0.0

        # How much of the correction to apply at all: the sights blending in,
        # and a melee attack committing, both come through here.
        self.tracking = 0.0

    # -- input ---------------------------------------------------------------

    def set_aim_direction(self, direction, facing_yaw):
        """Aim along a world-space direction, from a character facing this way.

        `direction` is in the game's coordinates -- y up, yaw 0 along +z -- so
        it can come from the camera's ray or from a locked-on target's position
        without either knowing about the other. `facing_yaw` is the s16 the
        character's feet are pointing along, which is what makes the result
        local: the same crosshair means a different twist depending on which
        way he is standing.
        """
        x, y, z = direction
        flat = math.hypot(x, z)
        self.target_yaw = s16_to_degrees(s16(atan2s(z, x) - facing_yaw))
        # Positive is up. Straight up is 90 and straight down is -90, both of
        # which the clamp below has opinions about.
        self.target_pitch = math.degrees(math.atan2(y, flat)) if flat or y else 0.0

    def set_tracking(self, strength):
        self.tracking = min(max(float(strength), 0.0), 1.0)

    # -- output --------------------------------------------------------------

    @property
    def local_yaw(self):
        """The clamped twist the torso is being asked for, in degrees."""
        limit = self.profile.yaw_limit
        return min(max(self.target_yaw, -limit), limit) * self.tracking

    @property
    def local_pitch(self):
        profile = self.profile
        wanted = self.target_pitch * profile.pitch_share
        return (min(max(wanted, profile.pitch_min), profile.pitch_max)
                * self.tracking)

    def twist_available(self, moving):
        """How much of the aim the torso is allowed to absorb, in degrees.

        Moving, the full twist the profile allows: his legs are busy carrying
        him somewhere and turning them would send him there sideways. Standing,
        only as much as is comfortable, so he squares up rather than staying
        wrung out at the waist indefinitely.

        None at all for a character with no pivot to twist -- Mario, whose
        skeleton tools/aim_rig.py has never been near. He aims by turning
        round, which is what the whole of this number being zero means.
        """
        if self.joint is None:
            return 0.0
        return self.profile.yaw_limit if moving else self.profile.comfort_yaw

    def body_turn(self, dt, moving):
        """How far the feet should come round this frame, in s16 units.

        docs/aim.md: the torso takes what it can and the character turns for
        the rest.

        Returns a signed s16 delta for the caller to apply to its own facing,
        or 0. Nothing here writes to the character.
        """
        profile = self.profile
        if self.tracking <= 0.0:
            return 0
        excess = abs(self.target_yaw) - self.twist_available(moving)
        if excess <= 0.0:
            return 0

        # Rate-limited by the excess itself, so the turn eases out as he comes
        # round rather than stopping dead the moment it is inside the limit.
        step = excess * min(profile.turn_rate * dt, 1.0) * self.tracking
        return degrees_to_s16(math.copysign(step, self.target_yaw))

    # -- per frame -----------------------------------------------------------

    def update(self, dt):
        """Spring the torso toward the aim and write it to the joint."""
        profile = self.profile
        self.yaw, self._yaw_velocity = _spring(
            self.yaw, self._yaw_velocity, self.local_yaw, profile.response, dt)
        self.pitch, self._pitch_velocity = _spring(
            self.pitch, self._pitch_velocity, self.local_pitch,
            profile.response, dt)

        if self.joint is not None:
            # Heading is the character's own yaw sense: the joint sits under a
            # node the game has already turned to his facing, so a positive
            # heading here is a positive turn there. Pitch leans him back,
            # which is what aiming up does to a body pivoting at the hips.
            self.joint.set_hpr(self.yaw, self.pitch, 0.0)

    def reset(self):
        """Drop the twist. For a character being put somewhere else."""
        self.target_yaw = self.target_pitch = 0.0
        self.yaw = self.pitch = 0.0
        self._yaw_velocity = self._pitch_velocity = 0.0
        self.tracking = 0.0
        if self.joint is not None:
            self.joint.set_hpr(0.0, 0.0, 0.0)

File: src/aim.rs
//! Turning the Hero's upper body to face where he is aiming.
//!
//! This is the procedural layer `docs/aim.md` specifies and its "As Built"
//! section records as **not ported**: the design, the rig work and the tuned
//! numbers all survived the move from Panda3D, and the code that drove them
//! did not. The numbers below are lifted from that section verbatim, and they
//! are on console sliders for the reason the doc gives -- every one of them was
//! arrived at by moving a slider, not by reasoning, so the next person to
//! disagree with one needs the slider more than they need the constant.
//!
//! What it drives is a single bone. `tools/aim_rig.py` inserts `AIM_TORSO`
//! between the skeleton root and every joint above the hips, carrying no
//! keyframes of its own, so a rotation written here rides on top of whatever
//! clip is playing rather than fighting it. The thighs, pelvis, belt and sash
//! stay behind, which is what lets the chest turn while the legs run.
//!
//! The bone's rest rotation is identity and its parent is the scene root, so
//! its local axes *are* model space: +Y up, +Z forward, the same convention
//! `player::movement` turns the body in and `billboard::facing` aims quads in.
//! That is why the rotation below is composed in plain world-ish terms instead
//! of through some bone-local correction.
//!
//! Only the twist is here. `docs/aim.md` also wants the rotation distributed up
//! the spine, authored aim-offset poses and left-hand IK, and all three need
//! the Rigify DEF bones parented properly in `TheHero.blend` first -- see the
//! "As Built" note on why the exported skeleton is flat. This is the doc's own
//! "simple first implementation can use only AIM_TORSO".

use crate::{
    console::GameTuning,
    level::LevelData,
    player::{Controller, Player, FIXED_DT},
    weapon::Loadout,
};
use bevy::{
    ecs::{schedule::ScheduleConfigs, system::ScheduleSystem},
    prelude::*,
    transform::TransformSystems,
};

/// The name `tools/aim_rig.py` gives the pivot it inserts.
///
/// By name, because that is the only thing that survives a glTF export -- the
/// same reason `billboard::claim` and the animation clips are looked up by
/// name.
const PIVOT: &str = "AIM_TORSO";

/// How far the aim ray is followed looking for something to converge on.
///
/// Only used to turn the camera's direction into a *point*, so that a shot
/// leaving the muzzle -- which is off to the side of the camera, in the Hero's
/// hand -- travels toward what the crosshair is on rather than off parallel to
/// it. Past this the ray has hit nothing and the point is simply far away,
/// which converges close enough that the difference is invisible.
const CONVERGE: f32 = 400.0;

/// Where the player is aiming, and how much of it the torso has taken up.
///
/// One resource rather than a component, because there is one player and every
/// reader of this wants the same answer: the weapon that fires along it, the
/// bone that turns toward it, and the feet that come round when it runs out of
/// twist.
#[derive(Resource, Debug, Clone, Copy)]
pub struct Aim {
    /// Unit vector the shot travels along, in world space.
    pub direction: Vec3,
    /// Where the camera's ray lands: the level surface it hits, or a point far
    /// down it if it hits nothing. Shots are aimed at this rather than along
    /// [`Self::direction`] so they converge on the crosshair.
    pub point: Vec3,
    /// The eye the ray was cast from, kept so a shot can tell whether the
    /// muzzle is already past what the camera was looking at.
    pub eye: Vec3,
    /// Smoothed torso twist, in radians, relative to the way the body faces.
    pub yaw: f32,
    /// Smoothed torso pitch, in radians. Positive is aiming up.
    pub pitch: f32,
    /// Yaw the torso was not allowed to take up, for the feet to come round
    /// by. Signed the same way as [`Self::yaw`].
    pub excess: f32,
    /// Spring velocities behind the two smoothed angles.
    yaw_rate: f32,
    pitch_rate: f32,
}

impl Default for Aim {
    fn default() -> Self {
        Self {
            direction: Vec3::Z,
            point: Vec3::ZERO,
            eye: Vec3::ZERO,
            yaw: 0.0,
            pitch: 0.0,
            excess: 0.0,
            yaw_rate: 0.0,
            pitch_rate: 0.0,
        }
    }
}

impl Aim {
    /// An aim looking `direction` and converging on `point`.
    ///
    /// The spring state stays private -- it is this module's working, and
    /// nothing outside has any business setting a torso's angular velocity --
    /// so this is how a caller builds one to ask a question of. Test support:
    /// the running game only ever gets its aim from [`drive`].
    #[cfg(test)]
    pub fn at(direction: Vec3, point: Vec3) -> Self {
        Self {
            direction,
            point,
            ..Self::default()
        }
    }
}

/// The pivot bone, claimed from the name the exporter gave it.
#[derive(Component)]
pub struct AimTorso;

/// Tags the pivot as the Hero's scene arrives.
pub fn claim(mut commands: Commands, arrivals: Query<(Entity, &Name), Added<Name>>) {
    for (entity, name) in &arrivals {
        if name.as_str() == PIVOT {
            commands.entity(entity).insert(AimTorso);
        }
    }
}

/// Wraps an angle into -pi..pi.
///
/// Every comparison here is between two headings, and without this a turn
/// across the back of the character reads as a turn the long way round: 179
/// degrees to -179 is two degrees of movement and looks like 358.
pub fn wrap(angle: f32) -> f32 {
    let turn = std::f32::consts::TAU;
    let wrapped = (angle + std::f32::consts::PI).rem_euclid(turn);
    wrapped - std::f32::consts::PI
}

/// The heading of a direction, in the same terms `player::movement` turns the
/// body in.
pub fn heading(direction: Vec3) -> f32 {
    direction.x.atan2(direction.z)
}

/// One step of a critically damped spring, toward `target`.
///
/// Critically damped is what `docs/aim.md` asks for and it is the right choice
/// rather than an arbitrary one: an underdamped torso overshoots the target and
/// swings back, which on a gun reads as the shot wandering off the crosshair
/// after the stick has stopped moving. `response` is the time it takes to
/// substantially arrive, so it is a number that can be reasoned about on a
/// slider -- the doc's 0.12 s.
///
/// The rational approximation of the exponential is the standard one from Game
/// Programming Gems 4; it is stable at any step size, which matters because
/// this runs at the render rate and a stalled frame must not fling the torso.
pub fn spring(value: f32, rate: f32, target: f32, response: f32, dt: f32) -> (f32, f32) {
    if response <= 0.0 || dt <= 0.0 {
        return (target, 0.0);
    }
    let omega = 2.0 / response;
    let x = omega * dt;
    let decay = 1.0 / (1.0 + x + 0.48 * x * x + 0.235 * x * x * x);
    let change = value - target;
    let temp = (rate + omega * change) * dt;
    (
        target + (change + temp) * decay,
        (rate - omega * temp) * decay,
    )
}

/// Reads the camera into [`Aim`], and turns the pivot toward it.
///
/// Runs per rendered frame rather than per fixed step because it is a pose:
/// the body it sits on is interpolated between fixed steps by
/// `player::sync_visual`, and a torso that snapped thirty times a second on top
/// of a body that glides is exactly the judder that system exists to avoid.
#[allow(clippy::too_many_arguments)]
pub fn drive(
    time: Res<Time>,
    tuning: Res<GameTuning>,
    loadout: Res<Loadout>,
    level: Res<LevelData>,
    mut aim: ResMut<Aim>,
    camera: Query<&GlobalTransform, With<Camera3d>>,
    player: Query<&Transform, With<Player>>,
    mut pivot: Query<&mut Transform, (With<AimTorso>, Without<Player>)>,
) {
    let Ok(view) = camera.single() else {
        return;
    };
    let eye = view.translation();
    // The camera's own forward, which is where the crosshair is pointing --
    // `camera::update` finishes by looking at the player, so this ray runs from
    // the eye through him and out the far side.
    let direction = Vec3::from(view.forward()).normalize_or(Vec3::Z);
    aim.eye = eye;
    aim.direction = direction;
    aim.point = match level.surface_hit(eye, eye + direction * CONVERGE) {
        Some((hit, _)) => hit,
        None => eye + direction * CONVERGE,
    };

    let dt = time.delta_secs();
    // A holstered gun does not twist anything. The sword's clips carry their
    // own swing and the doc's melee layer is not ported, so with the blade out
    // this unwinds to neutral and leaves the animation alone.
    let armed = loadout.equipped.is_ranged();
    let (mut wanted_yaw, mut wanted_pitch) = (0.0, 0.0);
    if armed {
        if let Ok(body) = player.single() {
            let facing = body.rotation * Vec3::Z;
            wanted_yaw = wrap(heading(direction) - heading(facing));
            // A fraction of the shot's elevation rather than all of it: the
            // chest leans into a high shot, the arm does the rest. The doc's
            // 55%, and the reason the pivot alone does not look like a turret.
            wanted_pitch = direction.y.clamp(-1.0, 1.0).asin() * tuning.torso_pitch;
        }
    }
    let limit = tuning.torso_limit.to_radians();
    let clamped_yaw = wanted_yaw.clamp(-limit, limit);
    let clamped_pitch = wanted_pitch.clamp(
        -tuning.torso_pitch_down.to_radians(),
        tuning.torso_pitch_up.to_radians(),
    );
    (aim.yaw, aim.yaw_rate) = spring(
        aim.yaw,
        aim.yaw_rate,
        clamped_yaw,
        tuning.torso_response,
        dt,
    );
    (aim.pitch, aim.pitch_rate) = spring(
        aim.pitch,
        aim.pitch_rate,
        clamped_pitch,
        tuning.torso_response,
        dt,
    );
    // What the twist could not cover is what the feet owe. Recorded here and
    // spent in `turn_body`, which runs on the fixed step the body moves on.
    aim.excess = if armed { wanted_yaw - clamped_yaw } else { 0.0 };

    for mut transform in &mut pivot {
        // Positive pitch aims up, and a positive rotation about +X tips the
        // model's +Z forward *down*, so the pitch goes on negated.
        transform.rotation = Quat::from_rotation_y(aim.yaw) * Quat::from_rotation_x(-aim.pitch);
    }
}

/// Brings the feet round when the twist runs out, and settles him square when
/// he is standing still.
///
/// Two limits rather than one, which is the doc's point: sixty degrees is how
/// far the torso *can* go, twenty is how far it is comfortable staying. Past
/// sixty the body turns because it must; between twenty and sixty it turns only
/// when he is not otherwise busy, so a player strafing while aiming behind
/// himself keeps the twist and a player stood still slowly squares up to what
/// he is looking at.
///
/// Fixed step, and after `player::movement`, because the rotation it writes is
/// the simulation's own -- `sync_visual` interpolates it, and a body turned in
/// the render schedule would be turned again from the stale value next tick.
pub fn turn_body(
    tuning: Res<GameTuning>,
    loadout: Res<Loadout>,
    aim: Res<Aim>,
    mut player: Query<(&mut Transform, &Controller), With<Player>>,
) {
    if !loadout.equipped.is_ranged() {
        return;
    }
    let Ok((mut transform, ctrl)) = player.single_mut() else {
        return;
    };
    let moving = Vec3::new(ctrl.velocity.x, 0.0, ctrl.velocity.z).length() > 0.25;
    // Standing still he settles back to the comfortable twist; running, only
    // the hard limit moves him, because the legs already have a direction and
    // the doc is explicit that locomotion owns them.
    let owed = if moving {
        aim.excess
    } else {
        let comfort = tuning.torso_comfort.to_radians();
        let total = aim.excess + aim.yaw;
        total - total.clamp(-comfort, comfort)
    };
    if owed.abs() < 1e-4 {
        return;
    }
    let step = tuning.torso_turn_rate.to_radians() * FIXED_DT;
    let turn = owed.clamp(-step, step);
    transform.rotation = Quat::from_rotation_y(heading(transform.rotation * Vec3::Z) + turn);
}

/// Claiming and driving, in the window where a pose may be written.
///
/// After the animation player, or the clip overwrites the pivot a moment later
/// -- `AIM_TORSO` carries no curves of its own, so today nothing would, but a
/// gun clip added later will and this must not have to be rediscovered. Before
/// the transforms are propagated, or the turn is a frame late and, worse,
/// everything hanging off the torso -- the arm, the hand, the gun in it -- is
/// drawn from a transform that never saw it. Exactly the reasoning in
/// `billboard::systems`, and the same two constraints.
pub fn systems() -> ScheduleConfigs<ScheduleSystem> {
    (claim, drive)
        .chain()
        .after(bevy::animation::animate_targets)
        .before(TransformSystems::Propagate)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn wrapping_takes_the_short_way_round() {
        assert!((wrap(0.0)).abs() < 1e-6);
        // Just past the back, which is the case that goes wrong untreated.
        let short = wrap(179f32.to_radians() - (-179f32).to_radians());
        assert!(
            short.abs() < 3f32.to_radians(),
            "{} degrees the long way",
            short.to_degrees()
        );
        // Three half-turns is a half-turn, and -pi and pi are the same
        // heading -- so it is the magnitude that is pinned here, not the sign.
        assert!((wrap(std::f32::consts::PI * 3.0).abs() - std::f32::consts::PI).abs() < 1e-4);
        for angle in [-9.0, -3.0, 0.4, 3.0, 9.0, 100.0] {
            let wrapped = wrap(angle);
            assert!(
                wrapped.abs() <= std::f32::consts::PI + 1e-5,
                "{angle} wrapped to {wrapped}, outside -pi..pi"
            );
            // Same heading, just named once round.
            assert!((wrapped.sin() - angle.sin()).abs() < 1e-4);
            assert!((wrapped.cos() - angle.cos()).abs() < 1e-4);
        }
    }

    #[test]
    fn heading_matches_the_way_the_body_turns() {
        // `player::movement` turns to `wish.x.atan2(wish.z)`; a body rotated by
        // that heading must end up facing the direction it came from.
        for direction in [Vec3::Z, Vec3::X, Vec3::NEG_Z, Vec3::new(1.0, 0.0, -1.0)] {
            let direction = direction.normalize();
            let facing = Quat::from_rotation_y(heading(direction)) * Vec3::Z;
            assert!(
                facing.dot(direction) > 0.999,
                "{direction:?} came back as {facing:?}"
            );
        }
    }

    /// Critically damped: it arrives, and it does not come back past the
    /// target on the way. An overshoot here is the shot wandering off the
    /// crosshair after the stick has stopped.
    #[test]
    fn the_spring_settles_without_overshooting() {
        let (mut value, mut rate) = (0.0f32, 0.0f32);
        let target = 1.0;
        for _ in 0..240 {
            (value, rate) = spring(value, rate, target, 0.12, 1.0 / 60.0);
            assert!(value <= target + 1e-4, "overshot to {value}");
        }
        assert!((value - target).abs() < 1e-3, "settled at {value}");
    }

    /// And it substantially arrives within the response time it was given,
    /// which is the whole reason that number is a number a person can tune.
    #[test]
    fn the_spring_arrives_in_about_its_response_time() {
        let (mut value, mut rate) = (0.0f32, 0.0f32);
        let dt = 1.0 / 60.0;
        for _ in 0..(0.12 / dt) as usize {
            (value, rate) = spring(value, rate, 1.0, 0.12, dt);
        }
        assert!(value > 0.55, "only reached {value} in one response time");
    }

    /// A huge frame must not fling the torso past the target.
    #[test]
    fn the_spring_is_stable_across_a_stalled_frame() {
        let (value, _) = spring(0.0, 0.0, 1.0, 0.12, 5.0);
        assert!((0.0..=1.0 + 1e-4).contains(&value), "flung to {value}");
    }
}
